rust/kernel/init.rs

```diff
@@ -0,0 +1,827 @@
+        unsafe { init.__pinned_init(slot)? };
+        // SAFETY: all fields have been initialized
+        Ok(unsafe { Pin::new_unchecked(this.assume_init()) })
+    }
+
+    fn init<E>(init: impl Init<T, E>) -> error::Result<Self>
+    where
+        Error: From<E>,
+    {
+        let mut this = Box::try_new_uninit()?;
+        let slot = this.as_mut_ptr();
+        // SAFETY: when init errors/panics, slot will get deallocated but not dropped,
+        // slot is valid
+        unsafe { init.__init(slot)? };
+        // SAFETY: all fields have been initialized
+        Ok(unsafe { this.assume_init() })
+    }
+}
+/*
+impl<T> InPlaceInit<T> for Arc<T> {
+    fn pin_init<E>(init: impl PinInit<T, E>) -> error::Result<Pin<Self>>
+    where
+        Error: From<E>,
+    {
+        let mut this = UniqueArc::try_new_uninit()?;
+        let slot = this.as_mut_ptr();
+        // SAFETY: when init errors/panics, slot will get deallocated but not dropped,
+        // slot is valid and will not be moved because of the `Pin::new_unchecked`
+        unsafe { init.__pinned_init(slot)? };
+        // SAFETY: all fields have been initialized
+        Ok(unsafe { Pin::new_unchecked(this.assume_init().into()) })
+    }
+
+    fn init<E>(init: impl Init<T, E>) -> error::Result<Self>
+    where
+        Error: From<E>,
+    {
+        let mut this = UniqueArc::try_new_uninit()?;
+        let slot = this.as_mut_ptr();
+        // SAFETY: when init errors/panics, slot will get deallocated but not dropped,
+        // slot is valid
+        unsafe { init.__init(slot)? };
+        // SAFETY: all fields have been initialized
+        Ok(unsafe { this.assume_init().into() })
+    }
+}*/
+
+impl<T> InPlaceInit<T> for UniqueArc<T> {
+    fn pin_init<E>(init: impl PinInit<T, E>) -> error::Result<Pin<Self>>
+    where
+        Error: From<E>,
+    {
+        let mut this = UniqueArc::try_new_uninit()?;
+        let slot = this.as_mut_ptr();
+        // SAFETY: when init errors/panics, slot will get deallocated but not dropped,
+        // slot is valid and will not be moved because of the `Pin::new_unchecked`
+        unsafe { init.__pinned_init(slot)? };
+        // SAFETY: all fields have been initialized
+        Ok(unsafe { Pin::new_unchecked(this.assume_init()) })
+    }
+
+    fn init<E>(init: impl Init<T, E>) -> error::Result<Self>
+    where
+        Error: From<E>,
+    {
+        let mut this = UniqueArc::try_new_uninit()?;
+        let slot = this.as_mut_ptr();
+        // SAFETY: when init errors/panics, slot will get deallocated but not dropped,
+        // slot is valid
+        unsafe { init.__init(slot)? };
+        // SAFETY: all fields have been initialized
+        Ok(unsafe { this.assume_init() })
+    }
+}
+
+/// Marker trait for types that can be initialized by writing just zeroes.
+///
+/// # Safety
+/// The bit pattern consisting of only zeroes must be a valid bit pattern for the type.
```

Comment:
```suggestion
/// # Safety
///
/// The bit pattern consisting of only zeroes must be a valid bit pattern for the type.
```

(and similar elsehwere)